plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1118 @@
+                {
+                    pubObjPtr->updateTimeTouched();
+                }
+                else
+                {
+                    // Publisher for that set of brokers and topic does not exist; create one
+                    try
+                    {
+                        pubObjPtr = new Publisher(_brokers, _topic, _pollTimeout, traceLevel);
+                        cachedPublishers[key] = pubObjPtr;
+
+                        if (traceLevel > 4)
+                        {
+                            DBGLOG("Kafka: Created and cached new publisher object: %s @ %s", _topic.c_str(), _brokers.c_str());
+                        }
+                    }
+                    catch (...)
+                    {
+                        // Make sure we're unlocked before re-throwing
+                        lockPtr->unlock();
+                        throw;
+                    }
+                }
+
+                lockPtr->unlock();
+
+                if (!pubObjPtr)
+                {
+                    throw MakeStringException(-1, "Kafka: Unable to create publisher for brokers '%s' and topic '%s'", _brokers.c_str(), _topic.c_str());
+                }
+
+                return pubObjPtr;
+            }
+
+        private:
+
+            ObjMap      cachedPublishers;   //!< std::map of created Publisher object pointers
+            Mutex*      lockPtr;            //!< Mutex guarding modifications to cachedPublishers
+            __int32     traceLevel;         //!< The current logging level
+    } *publisherCache;
+
+    //--------------------------------------------------------------------------
+
+    /** @class  PublisherCacheExpirerObj
+     *          Class used to expire old publisher objects held within publisherCache
+     */
+    static class PublisherCacheExpirerObj : public Thread
+    {
+        public:
+
+            PublisherCacheExpirerObj()
+                :   Thread("Kafka::PublisherExpirer")
+            {
+                shouldRun = false;
+            }
+
+            virtual ~PublisherCacheExpirerObj()
+            {
+
+            }
+
+            virtual void start()
+            {
+                if (!isAlive())
+                {
+                    shouldRun = true;
+                    Thread::start();
+                }
+            }
+
+            virtual void stop()
+            {
+                if (isAlive())
+                {
+                    shouldRun = false;
+                    join();
+                }
+            }
+
+            virtual int run()
+            {
+                while (shouldRun)
+                {
+                    if (publisherCache)
+                    {
+                        publisherCache->expire();
+                    }
+
+                    usleep(1000);
+                }
+
+                return 0;
+            }
+
+        private:
+
+            bool        shouldRun;      //!< If true, we should execute our thread's main event loop
```

Comment:
This should either be volatile, or even better std::atomic.  Otherwise a compiler would be able to optimize it away (it probably wouldn't because of the virtual call..